### PR TITLE
Finalization of the upgrade scheme

### PR DESCRIPTION
**Pull request by StanislavBreadless**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: l1-contracts/contracts/bridge/L2WrappedBaseToken.sol [hunk 1]
```diff
@@ -5,7 +5,7 @@ pragma solidity 0.8.24;
 import {ERC20PermitUpgradeable} from "@openzeppelin/contracts-upgradeable-v4/token/ERC20/extensions/draft-ERC20PermitUpgradeable.sol";
 
 import {IL2WrappedBaseToken} from "./interfaces/IL2WrappedBaseToken.sol";
-import {IBridgedStandardToken} from "./interfaces/IBridgedStandardToken.sol";
```
Comment (StanislavBreadless) [inline l1-contracts/contracts/bridge/L2WrappedBaseToken.sol:8]: I rolled back previous changes as we decided not to upgrade L2WETH, but I will have to return those (I likely will upgrade it)

File: system-contracts/scripts/constants.ts [hunk 2]
```diff
@@ -211,7 +211,15 @@ export const SYSTEM_CONTRACTS: ISystemContracts = {
...
     address: "0x0000000000000000000000000000000000010005",
-    codeName: "L2MessageRouter",
+    codeName: "MessageRoot",
+    lang: Language.Solidity,
+  },
+  SloadContract: {
+    // This is explicitly a non-system-contract address.
+    // We do not use the same address as create2 factories on EVM, since
+    // this is a zkEVM create2 factory.
+    address: "0x0000000000000000000000000000000000010005",
```
Comment (kelemeno) [inline system-contracts/scripts/constants.ts:221]:
```suggestion
    address: "0x0000000000000000000000000000000000010006",
```

File: l1-contracts/contracts/state-transition/chain-deps/facets/Admin.sol [hunk 3]
```diff
@@ -126,10 +131,6 @@ contract AdminFacet is ZKChainBase, IAdmin {
 
     /// @inheritdoc IAdmin
     function setPubdataPricingMode(PubdataPricingMode _pricingMode) external onlyAdmin onlyL1 {
-        // Validium mode can be set only before the first batch is processed
```
Comment (kelemeno) [inline l1-contracts/contracts/state-transition/chain-deps/facets/Admin.sol:129]: Removing this will allow even permanent rollups to set the feeparams.pubdataPricingMode to Validium. This does not change DA, but it does change L1->L2 txs fees in Mailbox _deriveL2GasPrice , the pubdataPriceBaseToken is now not added to fullPubdataPriceBaseToken. 

This could make L1->L2 txs fail. 

Not a big issue, just trying to think it through

File: l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol [hunk 4]
```diff
@@ -423,9 +429,14 @@ contract ExecutorFacet is ZKChainBase, IExecutor {
         // Save root hash of L2 -> L1 logs tree
         s.l2LogsRootHashes[_storedBatch.batchNumber] = _storedBatch.l2LogsTreeRoot;
 
-        // Once the batch is executed, we include its message to the message root.
-        IMessageRoot messageRootContract = IBridgehub(s.bridgehub).messageRoot();
-        messageRootContract.addChainBatchRoot(s.chainId, currentBatchNumber, _storedBatch.l2LogsTreeRoot);
+        // During migration to the new protocol version, there will be a period when
```
Comment (kelemeno) [inline l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol:432]: I would probably create a function to avoid duplication, but it is also ok like this

File: l1-contracts/contracts/bridge/L1Nullifier.sol [hunk 5]
```diff
@@ -741,4 +741,37 @@ contract L1Nullifier is IL1Nullifier, ReentrancyGuard, Ownable2StepUpgradeable,
...
+    /// @dev Note, that we need to keep this interface, since the `L2AssetRouter`
+    /// will continue returning the previous address as the `l1SharedBridge`. The value
+    /// returned by it is used in the SDK for finalizing withdrawals.
+    /// @param _chainId The chain ID of the transaction to check
+    /// @param _l2BatchNumber The L2 batch number where the withdrawal was processed
+    /// @param _l2MessageIndex The position in the L2 logs Merkle tree of the l2Log that was sent with the message
+    /// @param _l2TxNumberInBatch The L2 transaction number in the batch, in which the log was sent
+    /// @param _message The L2 withdraw data, stored in an L2 -> L1 message
+    /// @param _merkleProof The Merkle proof of the inclusion L2 -> L1 message about withdrawal initialization
+    function finalizeWithdrawal(
```
Comment (kelemeno) [inline l1-contracts/contracts/bridge/L1Nullifier.sol:760]: This does a L1N -> L1AR -> L1N routing. Maybe it makes sense to call the finalizeDeposit directly here? But it is also ok like this, it might overall be less complexity

File: l1-contracts/contracts/state-transition/chain-deps/facets/Admin.sol (hunk 3 above)
Comment (StanislavBreadless) [reply l1-contracts/contracts/state-transition/chain-deps/facets/Admin.sol:129]: makes sense, agree

File: l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol (hunk 4 above)
Comment (StanislavBreadless) [reply l1-contracts/contracts/state-transition/chain-deps/facets/Executor.sol:432]: agree

File: l1-contracts/contracts/bridge/L1Nullifier.sol (hunk 5 above)
Comment (StanislavBreadless) [reply l1-contracts/contracts/bridge/L1Nullifier.sol:760]: agree

File: l1-contracts/contracts/upgrades/BytecodesSupplier.sol [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+contract BytecodesSupplier {
+    /// @notice Event emitted when a bytecode is published.
+    event BytecodePublished(bytes32 indexed bytecodeHash, bytes bytecode);
+
+    /// @notice Mapping of bytecode hashes to the block number when they were published.
+    mapping(bytes32 bytecodeHash => uint256 blockNumber) public publishingBlock;
+
+    /// @notice Publishes the bytecode hash and the bytecode itself.
+    /// @param _bytecode Bytecode to be published.
+    function publishBytecode(bytes calldata _bytecode) public {
```
Comment (kelemeno) [inline l1-contracts/contracts/upgrades/BytecodesSupplier.sol:23]: Just to make sure EthWatch picks this up?

File: system-contracts/contracts/L2GatewayUpgrade.sol [hunk 7]
```diff
@@ -42,37 +42,47 @@ contract L2GatewayUpgrade {
...
-                (fixedForceDeploymentsData.l2BridgedStandardERC20Impl)
-            );
-            SystemContractHelper.mimicCallWithPropagatedRevert(
-                address(upgradableBeacon),
-                fixedForceDeploymentsData.l2BridgedStandardERC20ProxyOwnerAddress,
-                beaconUpgradeData
+            forceUpgradeProxy(
+                address(IL2SharedBridgeLegacy(l2LegacyBridgeAddress).l2TokenBeacon()),
+                // We are sure that `impl` is deployed, since it is supposed to included
+                // as part of the "usual" force deployments array.
```
Comment (kelemeno) [inline system-contracts/contracts/L2GatewayUpgrade.sol:62]: The token beacon is no longer updated

File: l1-contracts/contracts/upgrades/BytecodesSupplier.sol (hunk 6 above)
Comment (StanislavBreadless) [reply l1-contracts/contracts/upgrades/BytecodesSupplier.sol:23]: The server does work with it (tested locally and briefly reviewed by Misha). The scheme works roughly like that:

- This contract is the place where we publish bytecodes. Just publishing does not make them known on L2.
- When we receive an upgrade, we see that there is a transaction with it, but it contains only hashes for factory deps to be marked as known prior to the execution of the upgrade. It also must contain default aa hash/bootloader hash in the array.
- So basically the bytecodes are marked as known only once the upgrade tx is executed. The contract in question mainly serve just as a place for the server to search for the actual preimages

File: system-contracts/contracts/L2GatewayUpgrade.sol (hunk 7 above)
Comment (StanislavBreadless) [reply system-contracts/contracts/L2GatewayUpgrade.sol:62]: Why?

Comment (kelemeno) [reply system-contracts/contracts/L2GatewayUpgrade.sol:62]: sorry

File: l1-contracts/contracts/common/libraries/L2ContractHelper.sol [hunk 8]
```diff
@@ -85,6 +85,35 @@ library L2ContractHelper {
...
     }
 
+    /// @notice Validate the bytecode format and calculate its hash.
+    /// @param _bytecode The bytecode to hash.
+    /// @return hashedBytecode The 32-byte hash of the bytecode.
+    /// Note: The function reverts the execution if the bytecode has non expected format:
+    /// - Bytecode bytes length is not a multiple of 32
+    /// - Bytecode bytes length is not less than 2^21 bytes (2^16 words)
+    /// - Bytecode words length is not odd
+    function hashL2BytecodeCalldata(bytes calldata _bytecode) internal pure returns (bytes32 hashedBytecode) {
```
Comment (kelemeno) [inline l1-contracts/contracts/common/libraries/L2ContractHelper.sol:95]: In the L2Contracts its the other way around: 
https://github.com/matter-labs/era-contracts/blob/5b8b0785e4f7bae0b6d09d2dcd3d2b324204c16d/l2-contracts/contracts/L2ContractHelper.sol#L204

Comment (StanislavBreadless) [reply l1-contracts/contracts/common/libraries/L2ContractHelper.sol:95]: okay, I'll make it more uniform

Comment (StanislavBreadless) [reply l1-contracts/contracts/common/libraries/L2ContractHelper.sol:95]: It turned out changing l2-contracts folder was easier, so I will change it

File: l1-contracts/contracts/state-transition/l2-deps/IL2GenesisUpgrade.sol [hunk 9]
```diff
@@ -15,10 +15,16 @@ struct ForceDeployment {
...
 }
 
+// solhint-disable-next-line gas-struct-packing
 struct ZKChainSpecificForceDeploymentsData {
     bytes32 baseTokenAssetId;
     address l2LegacySharedBridge;
-    address l2Weth;
+    address predeployedL2WethAddress;
+    address baseTokenL1Address;
+    /// @dev Some info about the base token, it is
```
Comment (kelemeno) [inline l1-contracts/contracts/state-transition/l2-deps/IL2GenesisUpgrade.sol:24]: If you already include this, can it be used to set the baseTokenName in L2BaseToken?

Comment (StanislavBreadless) [reply l1-contracts/contracts/state-transition/l2-deps/IL2GenesisUpgrade.sol:24]: Discussed internally, considering timeline restrictions, it may be better to not add new features now (even if small, it will require of quite some tests. Also those methods in `L2BaseToken` are marked as unstable and so don't necessarily need to return correct values)
